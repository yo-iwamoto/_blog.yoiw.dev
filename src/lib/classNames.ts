type Args = (string | Record<string, boolean>)[];

// ポップアップが大きくなると邪魔なので JSDoc は使用していない
// @example cn('h-10', { 'hidden': false, 'flex': true }) // => returns 'h-10 flex'
export const cn = (...args: Args): string => {
  const classNames: string[] = [];

  args.forEach((value) => {
    if (typeof value === 'string') {
      classNames.push(value);
      return;
    }

    for (const [className, condition] of Object.entries(value)) {
      if (condition) {
        classNames.push(className);
      }
    }
  });

  return classNames.join(' ');
};
