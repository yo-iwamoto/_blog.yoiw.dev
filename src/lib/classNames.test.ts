import { cn } from './classNames';

const testCases = [
  {
    input: 'a',
    expectedOut: 'a',
  },
  {
    input: ['a', 'b', 'c'],
    expectedOut: 'a b c',
  },
  {
    input: [
      {
        included: true,
        'not-included': false,
      },
    ],
    expectedOut: 'included',
  },
  {
    input: [
      'a',
      {
        included: true,
        'not-included': false,
      },
      'b',
    ],
    expectedOut: 'a included b',
  },
] as const;

describe('classNames', () => {
  it.each(testCases)('正しい文字列になること', ({ input, expectedOut }) => {
    const result = cn(...input);

    expect(result).toBe(expectedOut);
    expect(result).toMatchSnapshot();
  });
});
