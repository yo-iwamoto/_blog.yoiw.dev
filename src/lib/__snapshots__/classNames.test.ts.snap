// Vitest Snapshot v1

exports[`classNames > 正しい文字列になること 1`] = `"a"`;

exports[`classNames > 正しい文字列になること 2`] = `"a b c"`;

exports[`classNames > 正しい文字列になること 3`] = `"included"`;

exports[`classNames > 正しい文字列になること 4`] = `"a included b"`;
